vars/sfdxBuildPipeline.groovy

```diff
@@ -70,165 +70,176 @@ def call(Map parameters = [:]) {
+                            pushToOrg org
                         }
-                    }
-                    if (!skipApexTests) {
-                        stage("${org.name} test") {
-                            runApexTests(org: org, timeoutMinutes: apexTestsTimeoutMinutes, usePolling: apexTestsUsePooling)
+                        
+                        if (packagesAfterPushStage.size() > 0) {
+                            stage("${org.name} install after push") {
+                                for (def p in packagesAfterPushStage) {
+                                    installPackage(org: org, package: p)
+                                }
+                            }
                         }
-                    }
-                    if (afterTestStage) {
-                        stage("${org.name} after test") {
-                            afterTestStage org
+                        
+                        if (beforeTestStage) {
+                            stage("${org.name} before test") {
+                                beforeTestStage org
+                            }
+                        }
+                        if (!skipApexTests) {
+                            stage("${org.name} test") {
+                                runApexTests(org: org, timeoutMinutes: apexTestsTimeoutMinutes, usePolling: apexTestsUsePooling)
+                            }
+                        }
+                        if (afterTestStage) {
+                            stage("${org.name} after test") {
+                                afterTestStage org
+                            }
+                        }
+                    } finally {
+                        stage("${org.name} delete") {
+                            if (keepOrg) {
+                                // To allow diagnosis of failures
+                                echo "Keeping scratch org name ${org.name} username ${org.username} password ${org.password} url ${org.instanceUrl} orgId ${org.orgId}"
+                            } else {
+                                echo "Deleting scratch org name ${org.name}"
+                                deleteScratchOrg org
+                            }
                         }
                     }
-                } finally {
-                    stage("${org.name} delete") {
-                        if (keepOrg) {
-                            // To allow diagnosis of failures
-                            echo "Keeping scratch org name ${org.name} username ${org.username} password ${org.password} url ${org.instanceUrl} orgId ${org.orgId}"
+                }
+                
+                stage("publish") {
+                    echo "Publishing test results"
+                    junit keepLongStdio: true, testResults: 'tests/**/*-junit.xml'
+                }
+            } catch (error) {
+                currentBuild.result = 'FAILURE'
+                println("Error: ${error}")
```

Comment:
@fischerp: I don't see value-added with this catch, could we simply keep the try/finally instead?

Reply:
The value is that besides the internal error you get an explicit detail about the error, which helps troubleshooting:
E.g.: `Error: hudson.AbortException: Script sfdx force:package:install --targetusername fake@example.com --package "prod_name" --installationkey PWD --wait 15 --noprompt failed: status 1`
Maybe the line 193 could be removed, but I decided to enforce the status because there could be multiple processes running in parallel.

Reply:
Fair enough, I think that as long as there is a catch without rethrowing the execution we need to set the build result to failure (line 193).

I was under the impression that the error message was being surfaced when a non-caught exception happens with the added benefit of displaying the stage where the failure occurred, if we kept this and now have surfaced more info, that is great news.